File: mobile/src/main/java/treehou/se/habit/ui/homescreen/ControllerWidget.java
package treehou.se.habit.ui.homescreen;

import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.Context;
import android.view.View;
import android.widget.RemoteViews;

import io.realm.Realm;
import treehou.se.habit.R;
import treehou.se.habit.core.db.model.controller.ControllerDB;
import treehou.se.habit.ui.control.ControllerUtil;

/**
 * Implementation of App Widget functionality.
 * App Widget Configuration implemented in {@link ControllerWidgetConfigureActivity ControllerWidgetConfigureActivity}
 */
public class ControllerWidget extends AppWidgetProvider {

    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
        // There may be multiple widgets active, so update all of them
        for (int appWidgetId : appWidgetIds) {
            updateAppWidget(context, appWidgetManager, appWidgetId);
        }
    }

    @Override
    public void onDeleted(Context context, int[] appWidgetIds) {
        // When the user deletes the widget, delete the preference associated with it.
        for (int appWidgetId : appWidgetIds) {
            ControllerWidgetConfigureActivity.deletePref(context, appWidgetId);
        }
    }

    @Override
    public void onEnabled(Context context) {

    }

    @Override
    public void onDisabled(Context context) {
    }

    static void updateAppWidget(Context context, AppWidgetManager appWidgetManager,
                                int appWidgetId) {

        Realm realm = Realm.getDefaultInstance();

        long controllId = ControllerWidgetConfigureActivity.loadControllIdPref(context, appWidgetId);
        boolean showTitle = ControllerWidgetConfigureActivity.loadControllShowTitlePref(context, appWidgetId);

        ControllerDB controller = ControllerDB.load(realm, controllId);
        if(controller == null){
            RemoteViews views = new RemoteViews(context.getPackageName(), R.layout.error_widget);
            appWidgetManager.updateAppWidget(appWidgetId, views);
            return;
        }

        // Construct the RemoteViews object
        RemoteViews views = new RemoteViews(context.getPackageName(), R.layout.controller_widget);
        views.removeAllViews(R.id.lou_rows);

        views.setInt(R.id.lou_widget, "setBackgroundColor", controller.getColor());
        views.setInt(R.id.lou_rows, "setBackgroundColor", controller.getColor());

        views.setTextViewText(R.id.lbl_title, controller.getName());
        views.setViewVisibility(R.id.lbl_title, showTitle ? View.VISIBLE:View.GONE);

        ControllerUtil.drawRemoteController(context, views, controller);

        // Instruct the widget manager to update the widget
        appWidgetManager.updateAppWidget(appWidgetId, views);

        realm.close();
    }
}

File: mobile/src/main/java/treehou/se/habit/ui/homescreen/ControllerWidget.kt
package treehou.se.habit.ui.homescreen

import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.Context
import android.content.Intent
import android.view.View
import android.widget.RemoteViews

import javax.inject.Inject

import io.realm.Realm
import treehou.se.habit.HabitApplication
import treehou.se.habit.R
import treehou.se.habit.core.db.model.controller.ControllerDB
import treehou.se.habit.ui.control.ControllerUtil

/**
 * Implementation of App WidgetFactory functionality.
 * App WidgetFactory Configuration implemented in [ControllerWidgetConfigureActivity]
 */
class ControllerWidget : AppWidgetProvider() {

    @Inject lateinit var controllerUtil: ControllerUtil

    override fun onUpdate(context: Context, appWidgetManager: AppWidgetManager, appWidgetIds: IntArray) {
        // There may be multiple widgets active, so update all of them
        for (appWidgetId in appWidgetIds) {
            updateAppWidget(context, appWidgetManager, appWidgetId)
        }
    }

    override fun onDeleted(context: Context, appWidgetIds: IntArray) {
        // When the user deletes the widget, delete the preference associated with it.
        for (appWidgetId in appWidgetIds) {
            ControllerWidgetConfigureActivity.deletePref(context, appWidgetId)
        }
    }

    override fun onReceive(context: Context, intent: Intent) {
        (context.applicationContext as HabitApplication).component().inject(this)
        super.onReceive(context, intent)
    }

    override fun onEnabled(context: Context) {}

    override fun onDisabled(context: Context) {}

    internal fun updateAppWidget(context: Context, appWidgetManager: AppWidgetManager,
                                 appWidgetId: Int) {

        val realm = Realm.getDefaultInstance()

        val controllId = ControllerWidgetConfigureActivity.loadControllIdPref(context, appWidgetId)
        val showTitle = ControllerWidgetConfigureActivity.loadControllShowTitlePref(context, appWidgetId)

        val controller = ControllerDB.load(realm, controllId)
        if (controller == null) {
            val views = RemoteViews(context.packageName, R.layout.error_widget)
            appWidgetManager.updateAppWidget(appWidgetId, views)
            return
        }

        // Construct the RemoteViews object
        val views = RemoteViews(context.packageName, R.layout.controller_widget)
        views.removeAllViews(R.id.lou_rows)

        views.setInt(R.id.lou_widget, "setBackgroundColor", controller.color)
        views.setInt(R.id.lou_rows, "setBackgroundColor", controller.color)

        views.setTextViewText(R.id.lbl_title, controller.name)
        views.setViewVisibility(R.id.lbl_title, if (showTitle) View.VISIBLE else View.GONE)

        controllerUtil.drawRemoteController(views, controller)

        // Instruct the widget manager to update the widget
        appWidgetManager.updateAppWidget(appWidgetId, views)

        realm.close()
    }
}


